### ISSUE TITLE
Word embeddings

### ISSUE THREAD

**Issue by EmilHvitfeldt** [tidymodels/textrecipes#20]
I want to include word embedding steps in this package at some point.

Asked in https://github.com/tidymodels/textrecipes/issues/9 I'm going to answer here in the main issue.

I would like to have a step for each word embedding we implement. I want this approach because I fear that a general `step_embed` would end up with a lot of arguments as each embedding might need different arguments.

As of right now the https://github.com/dselivanov/text2vec package is already imported and it allows for both text2vec and Glove so it would be a nice place to start. Later we can look at other embeddings such as fasttext which would require additional dependencies.

We would like to have some default pre-trained embeddings (for sure), but in the future it would also be nice for some custom trained embeddings to be used. Some investigation on the correct format would have to be done.
The priority right now is to have the embeddings working, and then we can implement user-supplied embeddings after.

**Comment by kanishkamisra** [comment]
Awesome!

So far I have two major points to make:

## Representation Format
Before we discuss the different algorithms, what kind of a representation do we want to return? Right now, `recipes` and `textrecipes` have a structure where each row represents a document, and each column (apart from the document name or label) is a feature. 

If we do want to implement embeddings we would have to use something like Continuous Bag of Words method (which I think is a okay-ish but acceptable approximation). Basically we would take the average of the vectors (along the dimensions) of every word in the document (which can be accessed by `step_tokenize()` ) and the resultant object would be each document represented by the n-dimensional vector of averaged word embeddings. 

Now, if we just want to keep this step as more of a word representation, we'd just have to look at a single document, take all its words (or desired words using some steps from the package) and then produce word vector estimates (based on number of dimensions as given by the user). This would be a little step away from how recipes and textrecipes do things. 

## Pretrained vectors or Model?
As far as pretrained representations go, we could do two things: 
1. Ask the user to download the pretrained vector .bin/.vec files such as [here](http://nlp.stanford.edu/data/glove.6B.zip) and then load only the vectors for words in the given document.
2. Train an algorithm to produce the representation of the document. So for GloVe, we'd use text2vec's functions and then proceed. 

OR we could do both!

Let me know your thoughts :)

**Comment by EmilHvitfeldt** [comment]
I was thinking we could use the functions in `text2vec`

So an example of going from list of tokens-to-tibble would be

``` r
library(textrecipes)
#> Loading required package: recipes
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(text2vec)
library(tokenizers)

text_vector <- okc_text$essay0
text_tokens <- tokenize_words(text_vector)

word_dims <- function(tokens, n = 10, n_iter = 20) {
  it <- text2vec::itoken(tokens, ids = seq_along(tokens))
  v <- text2vec::create_vocabulary(it)
  dtm <- text2vec::create_dtm(it, text2vec::vocab_vectorizer(v))
  lda_model <- text2vec::LDA$new(n_topics = n)
  d <- lda_model$fit_transform(dtm, n_iter = n_iter)
  d <- tibble::as_tibble(d)
  d
}

word_dims(text_vector, 10)
#> INFO [2018-12-24 16:25:30] iter 10 loglikelihood = -558384.525
#> INFO [2018-12-24 16:25:30] iter 20 loglikelihood = -546392.339
#> # A tibble: 750 x 10
#>        V1     V2     V3     V4     V5     V6     V7     V8     V9    V10
#>     <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>
#>  1 0.144  0.126  0.0778 0.126  0.210  0.0958 0.108  0.0599 0.0120 0.0419
#>  2 0.143  0.143  0.0476 0.0476 0.0952 0      0.143  0      0.238  0.143 
#>  3 0.0682 0.116  0.0979 0.128  0.0801 0.139  0.110  0.125  0.0593 0.0772
#>  4 0.0563 0.296  0.155  0.0845 0.0282 0.0704 0.0563 0.0986 0.0563 0.0986
#>  5 0.151  0.0548 0      0.0411 0.247  0.0959 0.0959 0.0822 0.178  0.0548
#>  6 0.1    0.044  0.056  0.06   0.028  0.02   0.132  0.22   0.204  0.136 
#>  7 0.0709 0.125  0.0676 0.0304 0.135  0.209  0.0574 0.0946 0.166  0.0439
#>  8 0.10   0.10   0.0429 0.0643 0.0929 0.121  0.0857 0.0857 0.10   0.207 
#>  9 0.25   0      0      0      0      0      0.25   0.25   0      0.25  
#> 10 0.0537 0.0755 0.195  0.0537 0.0656 0.0656 0.139  0.115  0.0577 0.179 
#> # ... with 740 more rows
```

<sup>Created on 2018-12-24 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Where I have drawn inspiration from the implementation in `textfeatures`'s [word2vec](https://github.com/mkearney/textfeatures/blob/master/R/word2vec.R) method. This should be fairly easy to implement into a step.

**Comment by kanishkamisra** [comment]
Ooh this sounds interesting, and easy enough to implement into textrecipes! Although, I feel this is more of a `step_topic_model()` rather than word2vec, due to the reason that with word embeddings, the output is a word representation, for example, a matrix with row vectors as the embeddings of words. With the LDA topic model, we're just getting the topic-wise distribution of each document, which is still a document-feature representation, but it just doesn't seem to fit into the embedding based document representations, but it does sound very appropriate for something like `step_topic_distribution()` or `step_topic_model()`.

Now, to represent documents with just word embeddings, literature has found several ways:

As an example, with the sentence: *Ho Ho Ho! Merry Christmas!*, the tokens are: `Ho, Ho, Ho, Merry, Christmas`. Using a pretrained embedding model, say GloVe, We can get each word's embeddings as follows and then sum them along their dimensions to represent the sentence.

```{r}
library(wordVectors)
library(tidyverse)
library(tokenizers)

sentence <- "Ho Ho Ho! Merry Christmas!"

## Download glove vectors from http://nlp.stanford.edu/data/glove.6B.zip
glove <- read.vectors("~/pretrained_wordvectors/glove.6B/glove.6B.50d.txt", binary = F)
embedding <- function(document, vector_space = glove) {
  tokens <- tokenize_words(document, simplify = T)
  embeds <- glove[tokens,] %>%
    colSums() %>%
    t() %>%
    as_tibble()
  return(embeds)
}
embedding(sentence)
#> # A tibble: 1 x 50
#>      V1    V2    V3    V4    V5    V6      V7     V8    V9   V10   V11
#>   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>  <dbl> <dbl> <dbl> <dbl>
#> 1 0.305  3.22 -1.62  3.28 -1.21 -4.60 -0.0553 -0.602 -1.92 0.492 -2.69
#> # ... with 39 more variables: V12 <dbl>, V13 <dbl>, V14 <dbl>, V15 <dbl>,
#> #   V16 <dbl>, V17 <dbl>, V18 <dbl>, V19 <dbl>, V20 <dbl>, V21 <dbl>,
#> #   V22 <dbl>, V23 <dbl>, V24 <dbl>, V25 <dbl>, V26 <dbl>, V27 <dbl>,
#> #   V28 <dbl>, V29 <dbl>, V30 <dbl>, V31 <dbl>, V32 <dbl>, V33 <dbl>,
#> #   V34 <dbl>, V35 <dbl>, V36 <dbl>, V37 <dbl>, V38 <dbl>, V39 <dbl>,
#> #   V40 <dbl>, V41 <dbl>, V42 <dbl>, V43 <dbl>, V44 <dbl>, V45 <dbl>,
#> #   V46 <dbl>, V47 <dbl>, V48 <dbl>, V49 <dbl>, V50 <dbl>
```

While this is done by summing the vectors of all the words along dimensions, [literature](https://arxiv.org/abs/1607.00570) has shown good results by also using the Min, Max and Mean along the dimensions. It isn't as good as the representations learned by training RNNs on documents as sequences, but it's possibly the best we could do with pretrained vectors (only bottleneck would be to load or train the model). We could include the aggregation function as a parameter.

A more large scale example with training the `text2vec` GloVe with 20 vectors (for fast training) on essay0:

```
library(tidyverse)
library(text2vec)
library(tokenizers)
library(textrecipes)

data("okc_text")
text_vector <- okc_text$essay0
word_embedding <- function(corpus) {
  tokens <- tokenize_words(corpus, simplify = T)
  it = itoken(tokens, progressbar = FALSE)
  vocab <- create_vocabulary(it)
  vectorizer <- vocab_vectorizer(vocab)
  tcm <- create_tcm(it, vectorizer, skip_grams_window = 5L)
  glove <- GlobalVectors$new(word_vectors_size = 20, vocabulary = vocab, x_max = 10)
  wv_main = glove$fit_transform(tcm, n_iter = 10, convergence_tol = 0.01)
  
  result_vectors <- t(glove$components) + wv_main
  
  return(result_vectors)
}
glove_vectors <- word_embedding(text_vector)
#> INFO [2018-12-25 15:03:36] 2018-12-25 15:03:36 - epoch 1, expected cost 0.0798
#> INFO [2018-12-25 15:03:36] 2018-12-25 15:03:36 - epoch 2, expected cost 0.0506
#> INFO [2018-12-25 15:03:36] 2018-12-25 15:03:36 - epoch 3, expected cost 0.0428
#> INFO [2018-12-25 15:03:37] 2018-12-25 15:03:37 - epoch 4, expected cost 0.0386
#> INFO [2018-12-25 15:03:37] 2018-12-25 15:03:37 - epoch 5, expected cost 0.0357
#> INFO [2018-12-25 15:03:37] 2018-12-25 15:03:37 - epoch 6, expected cost 0.0335
#> INFO [2018-12-25 15:03:37] 2018-12-25 15:03:37 - epoch 7, expected cost 0.0317
#> INFO [2018-12-25 15:03:37] 2018-12-25 15:03:37 - epoch 8, expected cost 0.0302
#> INFO [2018-12-25 15:03:38] 2018-12-25 15:03:38 - epoch 9, expected cost 0.0289
#> INFO [2018-12-25 15:03:38] 2018-12-25 15:03:38 - epoch 10, expected cost 0.0278
embedding <- function(document, vector_space = glove) {
  tokens <- tokenize_words(document)
  embeds <- map_df(tokens, function(x) {
    vector_space[x,] %>%
      colSums() %>%
      t() %>%
      as_tibble()
  })
  return(embeds)
}
embedding(text_vector, vector_space = glove_vectors)
#> # A tibble: 750 x 20
#>         V1     V2      V3     V4      V5      V6     V7     V8      V9
#>      <dbl>  <dbl>   <dbl>  <dbl>   <dbl>   <dbl>  <dbl>  <dbl>   <dbl>
#>  1 -130.    26.7  -118.    -4.24 -20.8     8.21  124.    63.4  -133.  
#>  2  -20.7   -3.11  -21.5    8.45  -5.39    9.52   17.6    6.03  -28.5 
#>  3 -320.    54.8  -337.    37.1  -24.0    20.2   302.   121.   -376.  
#>  4  -73.3    2.75  -70.9   19.1   -0.214  -3.92   63.5   37.4   -84.6 
#>  5  -72.2   40.8   -72.8  -14.9  -14.5     1.38   81.1    9.85  -61.3 
#>  6 -250.    34.2  -228.    32.2   -9.00   39.0   221.    95.8  -265.  
#>  7 -290.    80.5  -282.   -10.1  -44.6    31.7   296.    92.2  -300.  
#>  8 -134.    17.8  -136.    11.6   -8.84    6.29  129.    53.9  -152.  
#>  9   -4.91   1.39   -1.96  -1.77  -0.865  -0.510   1.87   2.21   -1.73
#> 10 -450.    34.2  -382.    77.2  -18.7    57.2   352.   226.   -475.  
#> # ... with 740 more rows, and 11 more variables: V10 <dbl>, V11 <dbl>,
#> #   V12 <dbl>, V13 <dbl>, V14 <dbl>, V15 <dbl>, V16 <dbl>, V17 <dbl>,
#> #   V18 <dbl>, V19 <dbl>, V20 <dbl>
```

One problem for now is that this is sound for short text documents (like tweets?) because we're aggregating the words of the document, for long documents this would be way too abstract.

Please do let me know if this makes sense, and I can start implementing it! 

Merry Christmas :)

**Comment by jonthegeek** [comment]
I'm playing around with implementing the pretrained embeddings version of this, pretty much as described by @kanishkamisra. My intended use is to keep the whole tibble of pretrained embeddings, to use them on words that weren't in the training data to still (hopefully) be able to do something with the new terms.

I'm wondering what you would suggest doing to ensure reproducibility of the recipe. I *could* include the entire tibble of embeddings in the prepped recipe step. However, the two smallest GloVe embeddings from Stanford are 187MB and 347MB, for example, so they'd balloon the recipe considerably. Any thoughts on that? I could also just point to the file from which the embeddings were loaded, but that could be unstable if you save the prepped recipe and then try to use it in the future.

It'll be useful to nail down this best practice, because I'll likely do ~the same thing for RBERT.

**Comment by EmilHvitfeldt** [comment]
To ensure reproducibility I'm afraid we would have to keep the embedding in the recipe.


we can use [textdata](https://github.com/emilhvitfeldt/textdata) for convenient download and storage of the pre-training files if you want

**Comment by jonthegeek** [comment]
OK, yeah, I agree, I was just hoping there would be a trick I wasn't thinking of.

Awesome about textdata! I'll definitely take a look. We could probably move the BERT checkpoint stuff over there and standardize a method for tracking things. 

I'll be diving deep into all of this tomorrow, but I'll probably leave the download step as assumed to be sorted out to get things working for a demo.

**Comment by EmilHvitfeldt** [comment]
Sounds good. Don't worry about trying to write a textdata function. I can do that for you once we figure out what needs to be done.

**Comment by jonthegeek** [comment]
I want to grok how it works to make sure we can get behind it for RBERT, so I'll at least take a look. It looks pretty straightforward. 😁

**Comment by jonthegeek** [comment]
OK, now that there's a convenient way to get embeddings via textdata, I'll work on this tomorrow. I'm thinking the user would supply the embeddings (with a demo of how to do it with textdata) as an argument. Can't decide if a wrapper function when they have textdata available would be worth it, but I'll code the agnostic version in the morning.
